src/cpp/session/modules/SessionPackages.R

```diff
@@ -1821,19 +1821,52 @@ if (identical(as.character(Sys.info()["sysname"]), "Darwin") &&
 })
 
 .rs.addJsonRpcHandler("validate_cran_repo", function(url) {
+   
+   url <- .rs.completeUrl(.rs.appendSlashIfNeeded(url), "src/contrib/PACKAGES.gz")
    packagesFile <- tempfile(fileext = ".gz")
    
-   tryCatch({
-      download.file(
-         .rs.completeUrl(.rs.appendSlashIfNeeded(url), "src/contrib/PACKAGES.gz"),
-         packagesFile,
-         quiet = TRUE
-      )
-
-      .rs.scalar(TRUE)
-   }, error = function(e) {
-      .rs.scalar(FALSE)
-   })
+   # ensure warnings are not suppressed in this scope
+   op <- options(warn = 1L)
+   on.exit(options(op), add = TRUE)
+   
+   # execute download.file(), but capture warnings and errors
+   warnings <- list()
+   error <- NULL
+   
+   withCallingHandlers(
+      
+      tryCatch(
+         download.file(url, packagesFile, quiet = TRUE),
```

Comment:
I wonder if we should do something a little more lightweight here, e.g. perform a HEAD request, or something else to avoid downloading the whole (potentially large) `PACKAGES.gz` file. Not sure if it's worth it though.